tfhe/src/core_crypto/gpu/vec.rs

```diff
@@ -29,6 +30,33 @@ pub struct CudaVec<T: Numeric> {
     _phantom: PhantomData<T>,
 }
 
+impl<T: Numeric> Clone for CudaVec<T> {
+    fn clone(&self) -> Self {
+        let size = self.len as u64 * std::mem::size_of::<T>() as u64;
+        let mut cloned_vec = Vec::with_capacity(self.ptr.len());
+        for &index in self.gpu_indexes.iter() {
+            unsafe {
+                cuda_synchronize_device(index);
```

Comment:
I agree synchronization is needed before copy, but not at this point. 

Since all the entire multi-gpu logic happens only at cuda side and everything is synchronized before returning to rust, we should synchronize `self.gpu_indexes[0]` only and before this loop.

Reply:
I can't pass any stream to clone, that's why I'm using the same logic than for Drop. I can't tell what happens on which GPU so it's safer to synchronize all of them.

Reply:
I don't get it. The rust side takes the stream at `gpu_indexes[0]` as the main stream, and the CUDA-sided methods with multi-gpu logic will synchronize all devices before returning. We shouldn't have anything happening on other devices than the first.

Reply:
We can't prevent someone from doing something on gpu_indexes[1] on the Rust side, so in this kind of method (Drop, Clone) we should act on all pointers in the CudaVec. I don't think this will impact performance significantly anyway, the Clone method is only there because we need to have an implementation of Clone for CudaVec in the HL API. We can always call duplicate on a ptr in the CudaVec as we usually do.